### PR TITLE
Friendship connections implemented

### PR DESCRIPTION
**Pull request by javierbotero**
Through the right associations, we created friendship connections for the users, inside the model User.

### CODE REVIEW

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

HEllo

Thank you for applying the changes. 
However, there are some issues or bugs to fix

0. **`PREREQUISITE`**: Did you solve the [**MILESTONE 5**](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)?

1. By being logged in as **Bob Bar**, I sent a friendship request to **Alice Bar**. Here is the result
Screenshot at 2020-06-29 23:16:32
You could notice that there is no indication concerning the relationship between the two user once now that the friendship request has been sent. It would be nice to have something like **Waiting for answer** or whatever else.

2. There is no button for **rejecting** a received friendship request. Please add it aside of the `Accept Request` button.

3. **Alice Bar** accepts the friendship request sent by **Bob Bar** and here is the content of the `Friendship` table
```rb
2.7.0 :001 > Friendship.all
  Friendship Load (0.6ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 1, user_id: 2, friend_id: 1, confirmed: true, created_at: "2020-06-29 22:16:05", updated_at: "2020-06-29 22:22:03">]> 
2.7.0 :002 > 
```
The problem is that is contains only one row instead of two which is the main requirement of this milestone. I suggest you re-read the last review message https://github.com/Bekhzod96/ror-social-scaffold/pull/3#pullrequestreview-437843796. The solution was give by the last reviewer. Implements the code put there.

4. Please, in the user lists, kindly add a message indicating the existence of a friendship relationship between the current user and the listed user (of course, if there is one)

5. There are no tests. **PLEASE ADD UNIT AND INTEGRATION TESTS**

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: app/views/users/index.html.erb
Comment (shshamim63) [inline app/views/users/index.html.erb]: You should generate all the data that are needed to render on the page in your controller. Too much logic in view is a bad practice. You can generate @frindship in your users_controller index method. After that, you can use it without any problem and without violating the best practice rules.
image

File: app/views/users/index.html.erb
Comment (shshamim63) [inline app/views/users/index.html.erb]: You have also violated the rule mentioned above. There should not be any fancy logic in views.
image

File: app/views/users/index.html.erb
Comment (shshamim63) [inline app/views/users/index.html.erb]: Same issue mentioned in above

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## :recycle: Changes Required :recycle: 
___
Hi @Bekhzod96 and @javierbotero :wave: 
Awesome job :clap: , You have covered most of the changes asked by the previous reviewer. But it needs some final touch up to get the approval. The good news is they are not tough to cover and for you, it is an easy job. Let's resolve all.
___
### :spades: Mandatory Changes :spades: 
___
- You should not put too much logic in your view. All data should be manipulated in the controller. And in view, you should use them. For example, in your post partial there you are looking for posts that are created by current_user and current_user's friend. But how you handled in view is conflicting with the Rails convention. I have a way to resolve it.
in your posts_controller
```ruby
def index
  @post = post.new
  user_with_posts_to_show = current_user.friends
  user_with_posts_to_show << current_user
  @posts = Post.timeline(user_with_posts_to_show)
end
```
in the post model
```ruby
scope :timeline, ->(users) { where(user_id: users).order(created_at: :desc) }
```
And you can @posts in your view.
- You should update the tests. The tests you have generated are not enough. You have added some validation in your models but you are not testing them at all. Take a look at this [article](https://dev.to/dtfocus/unit-testing-with-rspec-model-spec-4gii)
___
### :clubs: Optional Changes :clubs: 
- There are a couple of more violations of the Rails convention in your generated view files. In your users index.html.erb it contains some of them. As the previous reviewer did not mention it(I am really sorry) I am marking it as optional.
___
Please resolve the above-mentioned issue and send another review request. If you need any help feel free to knock me.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request reviewed by Maria Eugenia Queme P. (TSE)

### Hello @javierbotero and @Bekhzod96 :balloon: :lollipop: 

### Thank you for working on the changes requested by the previous reviewers :medal_sports:  :medal_sports: also thank you for all the hard work you've put in building this project. Let's work on the following changes :construction: :construction_worker: :construction_worker_man: 

#### Behaviour 

- [x] PostsController#index displays posts in a timeline (their own posts + posts from their friends) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/timeline.png) from [Odin Project](https://www.theodinproject.com/courses/ruby-on-rails/lessons/final-project))

- When a friendship gets accepted the current user should see the post of his or her friend in the TIMELINE but currently is not working this way, so let's make it work. I see you already handled this in your previous version but the last reviewer gave advice using scopes but you needed to take the friends into account in the scope so you can either modify the scope or use the following hint :point_down: 

_Hint: you could use a method like the following in your model to get the post from the user and his/her friends_
```ruby
def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

#### Associations

- [ ] for some reason every time you create a user the user friends her/him self and creates 4 rows, let's fix this and user shouldn't be friends with him/her self

image

#### Test cases

- [ ] I see you have made unit test cases but you need to add integration test cases using Rspec + Capybara and at least test the following scenarios (no need to test users, likes or posts **only everything related to friendships**):

- Send friend request
- Accept friend request
- Reject friend request
- See user + friends post in the TIMELINE

#### No logic in views or partials

* [ ] View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb). There should be absolutely NO logic in views.

  > Views are only for displaying information that became available to it from its corresponding controller not generating it. **Resources**
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

- The previous reviewer requested this and you've removed some but I think you haven't achieved to remove it all yet. there should be no unless, if, if/else, or if/end statements in your views for this you have to create helper methods that render the markup within partials, is not necessary you fix the logic that came with the repo but all the new code you added should follow best practices so only fix it in the views where you added code.

## [OPTIONAL]
#### Behaviour Milestone 5

- [ ] There is a button to invite the user to friendship on users list page and on the single-user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

- I see you can invite friends from the ALL USERS page but there should also be a button to invite a user in the user's profile (show page), I'm sorry this wasn't pointed out in the last milestone but this is a project requirement so I strongly recommend adding it into your project

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### Good job for making the changes from the previous reviews :clap: The application looks good and it has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one!
- Happy coding!

**Pull request merged**
